src/tool_help.c

```diff
@@ -35,6 +35,76 @@
 #  define USE_WATT32
 #endif
 
+/*
+ * The bitmask output is generated with the following command
+ ------------------------------------------------------------
+  cd $srcroot/docs/cmdline-opts
+  ./gen.pl listcats *.d
+ */
+
+#define CURLHELP_HIDDEN 1u << 0u
+#define CURLHELP_AUTH 1u << 1u
+#define CURLHELP_CONNECTION 1u << 2u
+#define CURLHELP_CURL 1u << 3u
+#define CURLHELP_DNS 1u << 4u
+#define CURLHELP_FILE 1u << 5u
+#define CURLHELP_FTP 1u << 6u
+#define CURLHELP_HTTP 1u << 7u
+#define CURLHELP_IMAP 1u << 8u
+#define CURLHELP_IMPORTANT 1u << 9u
+#define CURLHELP_MISC 1u << 10u
+#define CURLHELP_OUTPUT 1u << 11u
+#define CURLHELP_POP3 1u << 12u
+#define CURLHELP_POST 1u << 13u
+#define CURLHELP_PROXY 1u << 14u
+#define CURLHELP_SCP 1u << 15u
+#define CURLHELP_SFTP 1u << 16u
+#define CURLHELP_SMTP 1u << 17u
+#define CURLHELP_SSH 1u << 18u
+#define CURLHELP_TELNET 1u << 19u
+#define CURLHELP_TFTP 1u << 20u
+#define CURLHELP_TLS 1u << 21u
+#define CURLHELP_UPLOAD 1u << 22u
+#define CURLHELP_VERBOSE 1u << 23u
+
+typedef unsigned int curlhelp_t;
+
+struct category_descriptors {
+  const char *opt;
+  const char *desc;
+  curlhelp_t category;
+};
+
+static const struct category_descriptors categories[] = {
+  {"auth", "Options related to authentication of any kind", CURLHELP_AUTH},
+  {"connection", "Options related to low level network operations",
+   CURLHELP_CONNECTION},
+  {"curl", "Options related to the behaviour of curl itself", CURLHELP_CURL},
+  {"dns", "Options related to DNS", CURLHELP_DNS},
+  {"file", "Options related to the FILE protocol", CURLHELP_FILE},
+  {"ftp", "Options related to the FTP protocol", CURLHELP_FTP},
+  {"http", "Options related to the HTTP and HTTPS protocols", CURLHELP_HTTP},
+  {"imap", "Options related to the IMAP protocol", CURLHELP_IMAP},
+  /* important is left out because it is the default help page */
+  {"misc", "Options that don't fit in any other category", CURLHELP_MISC},
+  {"output", "Options that handle the output of the downloaded content",
+   CURLHELP_OUTPUT},
+  {"pop3", "Options related to the POP3 protocol", CURLHELP_POP3},
+  {"post", "Optons related to HTTP Post", CURLHELP_POST},
+  {"proxy", "Options related to proxy settings of any kind", CURLHELP_PROXY},
+  {"scp", "Options related to the SCP protocol", CURLHELP_SCP},
+  {"sftp", "Options related to the SFTP protocol", CURLHELP_SFTP},
+  {"smtp", "Options related to the SMTP protocol", CURLHELP_SMTP},
+  {"ssh", "Options related to the SSH protocol", CURLHELP_SSH},
+  {"telnet", "Options related to the Telnet protocol", CURLHELP_TELNET},
+  {"tftp", "Options related to the TFTP protocol", CURLHELP_TFTP},
+  {"tls", "Options related to any kind of TLS/SSL operation", CURLHELP_TLS},
+  {"upload", "Options related to uploads of any kind",
+   CURLHELP_UPLOAD},
+  {"verbose", "Options related to any kind of command line output of curl",
+   CURLHELP_VERBOSE}
+};
+
```

Comment:
Do you think we can remove the "Options related to" from each description? Would make every much more condensed, and to me it feels like redundant information. Perhaps instead add columns headers?